crates/metal-llvm/src/codegen.rs

```diff
@@ -32,66 +19,92 @@ pub struct CodeGen {
-            metal_ast::Primitives::I32 => unsafe { LLVMIntTypeInContext(self.ctx, 32) },
-            metal_ast::Primitives::I64 => unsafe { LLVMIntTypeInContext(self.ctx, 64) },
-            metal_ast::Primitives::I128 => unsafe { LLVMIntTypeInContext(self.ctx, 128) },
-
-            metal_ast::Primitives::F32 => unsafe { LLVMFloatTypeInContext(self.ctx) },
-            metal_ast::Primitives::F64 => unsafe { LLVMDoubleTypeInContext(self.ctx) },
-            metal_ast::Primitives::F128 => unsafe { LLVMFP128TypeInContext(self.ctx) },
-
-            metal_ast::Primitives::String => {
-                todo!()
-            }
-            metal_ast::Primitives::Void => unsafe { LLVMVoidTypeInContext(self.ctx) },
+    pub fn get_type(&self, t: metal_mir::types::Type) -> LLVMTypeRef {
+        match t {
+            metal_mir::types::Type::Primitive(p) => unsafe {
+                self.get_llvm_type_primitive(*(p.clone()))
+            },
+            metal_mir::types::Type::Composite(c) => unsafe {
+                match *c {
+                    metal_mir::types::Composite::Tuple(t) => LLVMStructTypeInContext(
+                        self.ctx,
+                        self.get_types(t.types.clone()).as_mut_ptr(),
+                        t.types.len().try_into().unwrap(),
+                        0,
+                    ),
+                    metal_mir::types::Composite::Array(a) => {
+                        LLVMArrayType2(self.get_type(a.item_type), a.size)
+                    }
+                }
+            },
+            metal_mir::types::Type::Function(f) => unsafe {
+                LLVMFunctionType(
+                    self.get_type(f.return_type),
+                    self.get_types(f.arguments.clone()).as_mut_ptr(),
+                    f.arguments.len().try_into().unwrap(),
+                    0,
+                )
+            },
         }
     }
 
-    pub fn ty(&self, ty: &metal_ast::Ty) -> LLVMTypeRef {
-        match ty {
-            metal_ast::Ty::Primitive(primitive) => self.primitive(primitive),
-            _ => {
-                todo!()
+    /// # Safety
+    /// This just calls a bunch of LLVM functions to return
+    /// a function.
+    pub unsafe fn get_llvm_type_primitive(
+        &self,
+        p: metal_mir::types::primitives::Primitive,
+    ) -> LLVMTypeRef {
+        match p {
+            // Integers (signed)
+            metal_mir::types::primitives::Primitive::I8 => LLVMIntTypeInContext(self.ctx, 8),
+            metal_mir::types::primitives::Primitive::I16 => LLVMIntTypeInContext(self.ctx, 16),
+            metal_mir::types::primitives::Primitive::I32 => LLVMIntTypeInContext(self.ctx, 32),
+            metal_mir::types::primitives::Primitive::I64 => LLVMIntTypeInContext(self.ctx, 64),
+            metal_mir::types::primitives::Primitive::I128 => LLVMIntTypeInContext(self.ctx, 128),
+
+            // Integers (unsigned)
```

Comment:
i believe these can be merged with the signed ones, e.g. `Primitive::I8 | Primitive::U8 => /* do the same stuff for both */`